// @miniflare/sites/dist/src/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
var __decorateClass = (decorators, target, key, kind) => {
  var result = kind > 1 ? void 0 : kind ? __getOwnPropDesc(target, key) : target;
  for (var i = decorators.length - 1, decorator; i >= 0; i--)
    if (decorator = decorators[i])
      result = (kind ? decorator(target, key, result) : decorator(result)) || result;
  if (kind && result)
    __defProp(target, key, result);
  return result;
};

// packages/sites/src/index.ts
__export(exports, {
  FilteredKVNamespace: () => FilteredKVNamespace,
  SitesPlugin: () => SitesPlugin
});

// packages/sites/src/filtered.ts
var import_kv = __toModule(require("@miniflare/kv"));
var FilteredKVNamespace = class extends import_kv.KVNamespace {
  #options;
  constructor(storage, options = {}, internalOptions) {
    super(storage, internalOptions);
    this.#options = options;
  }
  #included(key) {
    const options = this.#options;
    if (options.include !== void 0)
      return options.include.test(key);
    if (options.exclude !== void 0)
      return !options.exclude.test(key);
    return true;
  }
  get(key, options) {
    if (!this.#included(key))
      return Promise.resolve(null);
    return super.get(key, options);
  }
  getWithMetadata(key, options) {
    if (!this.#included(key)) {
      return Promise.resolve({ value: null, metadata: null });
    }
    return super.getWithMetadata(key, options);
  }
  async put(key, value, options) {
    if (this.#options.readOnly) {
      throw new TypeError("Unable to put into read-only namespace");
    }
    return super.put(key, value, options);
  }
  async delete(key) {
    if (this.#options.readOnly) {
      throw new TypeError("Unable to delete from read-only namespace");
    }
    return super.delete(key);
  }
  async list(options) {
    const { keys, list_complete, cursor } = await super.list(options);
    return {
      keys: keys.filter((key) => this.#included(key.name)),
      list_complete,
      cursor
    };
  }
};

// packages/sites/src/plugin.ts
var import_path = __toModule(require("path"));
var import_shared = __toModule(require("@miniflare/shared"));
var SitesPlugin = class extends import_shared.Plugin {
  sitePath;
  siteInclude;
  siteExclude;
  #setupResult;
  constructor(ctx, options) {
    super(ctx);
    this.assignOptions(options);
    this.#setupResult = this.#setup();
  }
  async #setup() {
    if (!this.sitePath)
      return {};
    const {
      FileStorage
    } = require("@miniflare/storage-file");
    const sitePath = import_path.default.resolve(this.ctx.rootPath, this.sitePath);
    const storage = new FileStorage(sitePath, false);
    const bindings = {
      __STATIC_CONTENT: new FilteredKVNamespace(storage, {
        readOnly: true,
        include: this.siteInclude && (0, import_shared.globsToMatcher)(this.siteInclude),
        exclude: this.siteExclude && (0, import_shared.globsToMatcher)(this.siteExclude)
      }),
      __STATIC_CONTENT_MANIFEST: {}
    };
    const additionalModules = {
      __STATIC_CONTENT_MANIFEST: { default: "{}" }
    };
    return { bindings, watch: [sitePath], additionalModules };
  }
  async setup() {
    return this.#setupResult;
  }
};
__decorateClass([
  (0, import_shared.Option)({
    type: import_shared.OptionType.STRING,
    name: "site",
    alias: "s",
    description: "Path to serve Workers Site files from",
    logName: "Workers Site Path",
    fromWrangler: ({ site }) => site?.bucket
  })
], SitesPlugin.prototype, "sitePath", 2);
__decorateClass([
  (0, import_shared.Option)({
    type: import_shared.OptionType.ARRAY,
    description: "Glob pattern of site files to serve",
    logName: "Workers Site Include",
    fromWrangler: ({ site }) => site?.include
  })
], SitesPlugin.prototype, "siteInclude", 2);
__decorateClass([
  (0, import_shared.Option)({
    type: import_shared.OptionType.ARRAY,
    description: "Glob pattern of site files not to serve",
    logName: "Workers Site Exclude",
    fromWrangler: ({ site }) => site?.exclude
  })
], SitesPlugin.prototype, "siteExclude", 2);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FilteredKVNamespace,
  SitesPlugin
});
